Revit_Core_Engine/Query/CloudGuids.cs

```diff
@@ -0,0 +1,56 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2024, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using Autodesk.Revit.DB;
+using BH.oM.Base;
+using BH.oM.Base.Attributes;
+using System;
+using System.ComponentModel;
+
+namespace BH.Revit.Engine.Core
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Returns project and model guids of a model hosted in the cloud. Null if the model is not hosted in the Autodesk cloud.")]
+        [Input("document", "Cloud model to query for guids.")]
+        [MultiOutput(0, "projectGuid", "Project guid of a model hosted in the cloud.")]
+        [MultiOutput(1, "modelGuid", "Model guid of a model hosted in the cloud.")]
+        public static Output<Guid, Guid> CloudGuids(this Document document)
```

Comment:
Hm, interesting, I would never think of returning guids of all elements 🙈 The suggested name is a bit lengthy, maybe `CloudGuidsOfModel` would be a good meeting point?

Then on the code suggestions, not sure if I understand, why would we want to try return cloud guids of a non-cloud model? What sort of guids would you return from non-cloud models?

Reply:
From what I understand, the committed code does a few things: 
1. Takes in a document
2. Uses that document to get a ModelPath
3. Uses that ModelPath to get the ProjectGuid and ModelGuid

If its not possible for a non-cloud hosted document to have Guids, then from what I see, its really the ModelPath that is providing the Guids.

It seems possible that from the ModelPath, it could be a file stored on disk, server, or cloud
https://www.revitapidocs.com/2023/f214cb44-8350-054e-821a-53571769f236.htm
![image](https://github.com/BHoM/Revit_Toolkit/assets/96432552/fc6adc0b-6e3c-487c-8545-60b5a83ebd74)

I'm thinking that the abstraction starts at the wrong class, and that if we are intending to specifically only get Guids from a known cloud hosted document, there would be more checks around ModelPath class rather than on the document instead. You possibly can also get rid of the try catch this way. 


```
//using your proposed method, but modifying the method signature. 

if(modelPath.CloudPath) return modelPath.CloudGuids();

```

```suggestion
        public static Output<Guid, Guid> CloudGuids(this ModelPath modelPath)
```

Reply:
I am not sure if we understand each other in terms of the intent of this method @travispotterBH. As mentioned in the title of the related issue, we need a wrapper method that would return cloud guids of a document - in other words, I want to have a method that will return cloud guids of a model if it is hosted in the cloud to avoid boilerplate related to checking if the model is hosted in the cloud and calling `Document.GetCloudModelPath` in multiple places.

Reply:
I'm not sure how different it is to check to see if an element is a cloud model before  versus then needing to check for null on the return of this method. But I do feel relying on null as a return is an anti-pattern. I know we rely on null as return a lot in the BHoM, I supposed I'm advocating for slightly less use of it. 

```
if(!document.IsModelInCloud()) //do something else

var guids = document.ModelPath.CloudGuids()

```
versus

```
var guids = document.CloudGuids();
if(guids == null){
//error? 
//other stuff? 
}

Reply:
Well, the difference is very simple:
- in my proposal you can either check if the model is a cloud model before calling `doc.CloudGuids()` or check against `null` afterwards - up to you as a developer
- in your suggestion you force the consumer to remember to check if the model is a cloud model, otherwise you end up with an exception

I appreciate you strong opinion about the patterns we use in BHoM, but I still am an advocate of consistency and cognitive freedom.